### ISSUE TITLE
Dam Observation Reach Assignment

### ISSUE THREAD

**Issue by philipbaileynar** [Riverscapes/sqlBRAT#52]
One of the last tasks to implement in sqlBRAT is assigning actual dam observation data to reaches. I am curious how this was done in pyBRAT @wally-mac and @tyler1218hatch ?

I have all the observation data for each HUC stored in the BRAT database along with the reaches and I'm ready to assign each dam observation to a reach. My plan is to loop over all the observations in a watershed and buffer them by some nominal amount, say 5m, and then look for reaches within this buffer polygon. I will repeat this process, incrementally growing the buffer by some nominal increment until the nearest reach is found.

If multiple reaches are found in any particular buffer then I can resolve using a distance calculation.

I think this will work. It should also be fairly efficient because there are typically few observations relative to the number of reaches within a watershed.

@wally-mac , @tyler1218hatch @MattReimer any thoughts on this?

**Comment by MattReimer** [comment]
Another method would be to set a maximum distance (i.e. anything greater than 1km is not useful) and just do geometry intersections on the small-ish number of objects inside the circle.

Unless there's some metadata on the dam observations we can use to link it.

**Comment by wally-mac** [comment]
@MattReimer @philipbaileynar we have a well documented approach for dealing with dam census data and the validation process. I want to get ﻿ @tyler1218hatch ﻿ up to speed on this and have him provide you with the necessary documentation on how this is performed. Thanks!

**Comment by tyler1218hatch** [comment]
@philipbaileynar @wally-mac I'll look into this right now and give you a write up on how this is performed!

**Comment by tyler1218hatch** [comment]
@philipbaileynar  @wally-mac Most of my info is from the [BRAT documentation](http://brat.riverscapes.xyz/Documentation/Tutorials/9-DataValidation.html) and the [code itself](https://github.com/Riverscapes/pyBRAT/blob/master/Data_Capture_Validation.py).

The first step is to determine if you want to use a drainage area threshold. 

**If no threshold is provided**, the dams are simply snapped to the nearest segment within 60 meters (Data_Capture_Validation.py [146])

**If a drainage area threshold is provided,**

1.	Create a new shapefile with only segments above the threshold `(Data_Capture_Validation.py [139])`
2.	Try to snap all dams to this shapefile, but only within 5 meters `(Data_Capture_Validation.py [143])`
3.	Create a new shapefile with only segments below the threshold `(Data_Capture_Validation.py [141])`
4.	Try to snap all dams to this shapefile, within 60 meters `(Data_Capture_Validation.py [144])`

_(Interestingly, I think I found a bug in the old code at Data_Capture_Validation.py [137 & 138]. These lines should end with da_threshold, not 65.)_

The reason behind using a drainage area threshold is that we have found some of these dams to be located on side channels which are not represented in the NHD network. Limiting snapping on streams above the drainage threshold thus helps to minimize the chance that dams will be snapped to a reach where they do not actually occur. Dams are a lot less likely to appear on reaches with a very high drainage area, so we can only assume that the dam was actually on the segment if the point is very close (5 meters).

This process will inevitably leave some dams unsnapped, with no NHD segment that they correspond to. So, we create a field named “Snapped” to distinguish between snapped and unsnapped dams. As I understand it, the unsnapped dams aren’t used for much after this. All future steps will only occur with dams that are snapped onto the network. Now we add all of the fields that are important to segments with observed dams.

•	**e_DamCt**: Simply the number of observed dams on this reach. This value will typically be 0 for most reaches. `(Data_Capture_Validation.py [170])`

•	**e_DamDens**: The number of dams per kilometer. Calculated by dividing e_DamCt by the segment length. `(Data_Capture_Validation.py [171])`

•	**e_DamPcC**: The ratio between the observed dam count and the calculated existing capacity. Calculated by dividing e_DamCt by oCC_EX. `(Data_Capture_Validation.py [179])`

•	**BRATvSurv**: The ratio between the estimated existing capacity count (mCC_EX_CT) and the surveyed dam capacity count (e_DamCt). `(Data_Capture_Validation.py [173])`

•	**ConsVRest**: A reclassification of the oPBCR_CR field from the conservation restoration model based on surveyed dam locations, with categories including: 

      o Immediate: Beaver conservation - reaches classified as “easiest -low-hanging fruit” with at least 25% of existing dam capacity already occupied by surveyed dams

      o Immediate: Potential Beaver Translocation - reaches classified as “easiest - low-hanging fruit” with less than 25% of existing dam capacity occupied by surveyed dams

      o Mid Term: Process-based Riparian Vegetation Restoration - reaches classified as “Straight Forward - Quick Return”

      o Long Term: Riparian Vegetation Reestablishment - reaches classified as “Strategic - Long-Term Investment”

      o Low Capacity Habitat - reaches classified as “NA”

      `(Data_Capture_Validation.py [185])`

In summary, if we don't want to use a drainage threshold, this should be relatively easy. Just use a snapping tool and set the distance to 60m. _(If a snapping tool doesn't work, then I suppose some other method using buffers and intersections would also work.)_ If we do want to use a drainage area threshold, this would have to be another parameter the user defines. I don't think there would be a single threshold that applies to all watersheds. If segments are above that threshold only snap the dam if it's within 5m. Otherwise snap normally.

Let me know what you all think! This process isn't concrete, so if it needs to be altered we can make that happen.
